rfcs/0133-ipfs.md

```diff
@@ -0,0 +1,284 @@
+NarSize: 197528
+References: 7gx4kiv5m0i7d7qkixq2cwzbr10lvxwc-glibc-2.27 gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
+Deriver: 5sj6fdfym58sdaf3r5p87v4l8sj2zlvn-hello-2.10.drv
+Sig: cache.nixos.org-1:K0thQEG60rzAK8ZS9f1whb7eRlIshlMDJAm7xvX1oF284H+PTqlicv/wGW6BIj+wWWONHvUZ2MYc+KDArekjDA==
+```
+
+This RFC proposes new key-value pairs that in this example would be:
+
+```
+IpfsCid: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN
+IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}
+```
+
+Just as today, the `NarHash` and `NarSize` remain the *normative* way to lock down the store object the `narinfo` file describes.
+Conversely, The `URL`, `FileHash` and `FileSize` by contrast are *informational*, describing not what the store object *is*, but *how to get it*.
+
+The `IpfsCid` and `IpfsEncoding` are likewise informational, describing how to get the store object:
+
+- `IpfsCid`: Native content address for IPFS.
+
+- `IpfsEncoding`: Enough info to deterministically rebuild the IPFS representation from a non-IPFS copy of the store object.
+
+   For now, `IpfsEncoding` will only support `unixfs-nar`, which works as follows:
+
+   The NAR is itself wrapped in IPFS's [UnixFS](https://github.com/ipfs/specs/blob/main/UNIXFS.md).
+   This other format can be extracted from the CID (which is conceptually a pair of encoding metadata and a hash).
+   For now, only IPFS's "unixfs" is supported.
+   `chunking`, `layout`, and `max-size` are tuning parameters for unixfs [described in the UnixFS spec](https://github.com/ipfs/specs/blob/main/UNIXFS.md#importing).
+
+   "UNIXFS" is not used directly because it doesn't support the "executable bit** Nix does on files.
+   NAR archive are not used directly because IPFS doesn't support arbitrary large objects.
+
+## IPFS Narinfo and "stateful" IPFS Store
+
+- **Purpose**: Binary distribution
+- **Depends on**: Augmented `narinfo`
+
+Instead of a "hybrid" store, where the narinfo index is served with HTTP but the data itself is served with IPFS, we can do an all-IPFS store with the data itself and mutable index stored in IPFS.
+The Narinfo instead of being encoded the legacy line-oriented text format can be IPFS's native DAG-CBOR IPLD codec, which is like JSON + content address links (but stored as CBOR).
+This allows Narinfos to reference each other and be nicely structured so the index is legible from Nix-agnostic IPFS tools and recursive pinning comes for free.
+
+Read-only is easier, since IPFS data is immutable but "writable" stores are supported by simple printing back a new CID for the new store root after some modifications, or modifying a mutable IPNS reference.
+IPNS is historically slow, but the update is automatic.
+Printing out a new CID for the index root allows the store administrator to update an out-of-bound mutable reference, but this cannot be automated because Nix doesn't know what the out-of-band method is.
```

Comment:
> but this cannot be automated because Nix doesn't know what the out-of-band method is.

I'm not sure the point here. Just as updating IPNS can be automated we can automate publishing DNSLink entries for popular providers or any number of other methods. Maybe you trying to point out that this would be a protocol outside of IPFS?

Reply:
Yes exactly. It is quite possible to get Nix to call some hook, but there are perhaps details to figure out there (though I suppose we've rather wantonly added hooks so far...) and I didn't do it yet.

There is an interesting thing going on where I am very keen on the source code use-case, which due to it's statelessness is just much easier / more ergonomic on both the "provider" and "consumer" sides, but it also requires more up-front changes to Nix which could be controversial.

I look forward to RFC meetings to sort out all that.

Reply:
It isn't as trendy, but Bittorrent's [ed25519-signed DHT entries](https://www.bittorrent.org/beps/bep_0044.html) can be used for this.  The protocol is very lightweight and easy to understand -- not to mention battle-tested.  Getting block-level content addressing requires [v2 torrents](http://www.bittorrent.org/beps/bep_0052.html) which are still not as popular as the v1 torrents -- but they share the same DHT network and uTP protocol.

Reply:
Yeah I am down to experiment the other protocols here. As I've long maintain, very little of the work here is *actually* specific to IPFS.

Reply:
Collapsing IPFS threads.